### PR TITLE
[WIP]Add type hinting

### PR DESCRIPTION
**Pull request by Tyagdit**
This could help us debug better and better deal with edge cases, not sure if what I've done here is good enough but it's a start

This is going to be a pretty large refactor so it would be better to be slow and thorough

## TODO:
- [ ] Add type checking to CI
- [ ] Add type checking to pre-commit hooks
- [ ] Update contributing

### CODE REVIEW

**Review by obviyus — APPROVED** [review]

File: requirements.txt [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
-pillow==8.1.0
-pre-commit>=2.9.3
-python-telegram-bot==13.1
-pyyaml==5.3.1
+googletrans==4.0.0-rc1
+gtts==2.2.2
+howlongtobeatpy==0.1.17
+mysql-connector-python==8.0.23
+pillow==8.1.2
+pre-commit==2.11.1
```
Comment (Tyagdit) [inline requirements.txt]: Why pin this? this is a dev dependency

File: requirements.txt [hunk 2]
```diff
@@ -1,19 +1,20 @@
 Babel==2.9.0
 cloudmersive_image_api_client==3.0.1
 dateparser==1.0.0
-emoji>=0.6.0
+emoji==1.2.0
```
Comment (Tyagdit) [inline requirements.txt]: dont think we need this pinned either

File: requirements.txt [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
+python-telegram-bot==13.3
+pyyaml==5.4.1
 qbittorrent-api==2020.12.15
 requests==2.25.1
 spurdify==0.3.0
-urllib3==1.26.2
-wolframalpha==4.1.1
+urllib3==1.26.3
+wolframalpha==5.0.0
+regex==2020.11.13
```
Comment (Tyagdit) [inline requirements.txt]: alphabetical bro

File: chat_management/kick.py [hunk 4]
```diff
@@ -1,7 +1,18 @@
...
-    message = update.message
-    chat_id = message.chat_id
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
+    chat_id: int = message.chat_id
+    text: str
 
     kicker = context.bot.get_chat_member(chat_id, message.from_user.id)
```
Comment (Tyagdit) [inline chat_management/kick.py]: missed annotation here

Comment (Tyagdit) [reply chat_management/kick.py]: also `user_to_kick` further down

File: chat_management/ban.py [hunk 5]
```diff
@@ -1,7 +1,18 @@
...
-    message = update.message
-    chat_id = message.chat_id
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
+    chat_id: int = message.chat_id
+    text: str
 
     banner = context.bot.get_chat_member(chat_id, message.from_user.id)
```
Comment (Tyagdit) [inline chat_management/ban.py]: same here

File: api/weather.py [hunk 6]
```diff
@@ -1,20 +1,15 @@
...
-    n = 2.0 ** zoom
-    xtile = int((lon_deg + 180.0) / 360.0 * n)
-    ytile = int((1.0 - math.asinh(math.tan(lat_rad)) / math.pi) / 2.0 * n)
-    return xtile, ytile
-
+if TYPE_CHECKING:
+    import telegram
+    import telegram.ext
 
 weather_codes = {
```
Comment (Tyagdit) [inline api/weather.py]: here too

Comment (Tyagdit) [reply api/weather.py]: bunch more further down as well

File: api/wait.py [hunk 7]
```diff
@@ -1,25 +1,37 @@
...
+    import telegram
+    import telegram.ext
+
+
+def wait(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -> None:
     """What Anime Is This?"""
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
```
Comment (Tyagdit) [inline api/wait.py]: why do this if you're reassigning immediately after this

File: api/wait.py [hunk 8]
```diff
@@ -1,25 +1,37 @@
...
+
+    text: str
     try:
-        message = update.message.reply_to_message
+        message = message.reply_to_message
 
         file = context.bot.getFile(message.photo[-1].file_id)
```
Comment (Tyagdit) [inline api/wait.py]: 3 more after this one

File: api/ud.py [hunk 9]
```diff
@@ -1,18 +1,28 @@
...
-        text = "*Usage:* `/ud {QUERY}`\n"\
+        text = "*Usage:* `/ud {QUERY}`\n" \
                "*Example:* `/ud boomer`\n"
     else:
-        result = get(f"http://api.urbandictionary.com/v0/define?term={word}")
-        result = result.json()
+        result: Dict = get(f"http://api.urbandictionary.com/v0/define?term={word}").json()
 
         if result['list']:
             # Sort to get result with most thumbs up
```
Comment (Tyagdit) [inline api/ud.py:28]: missed some in this part

File: api/tts.py [hunk 10]
```diff
@@ -1,16 +1,27 @@
...
+        message: 'telegram.Message' = update.message
+    else:
+        return
+
+    text: str
     if not context.args:
         try:
             sentence = message.reply_to_message.text or message.reply_to_message.caption
-            tts = gTTS(sentence, lang='ja')
+            speech = gTTS(sentence, lang='ja')
```
Comment (Tyagdit) [inline api/tts.py]: here too

File: api/translate.py [hunk 11]
```diff
@@ -1,25 +1,35 @@
...
-            text = translated.text
+            text = translated.__dict__()["text"]
         except AttributeError:
-            # blame Udit for this example
-            text = "*Usage:* `/tl {DEST} - {SENTENCE}`\n"\
-                   "*Example:* `/tl en - watashi wa anato no suki desu`\n"\
-                   "Defaults to `en` if none provided.\n"\
+            text = "*Usage:* `/tl {DEST} - {SENTENCE}`\n" \
+                   "*Example:* `/tl en - watashi wa anato no suki desu`\n" \
+                   "Defaults to `en` if none provided.\n" \
```
Comment (Tyagdit) [inline api/translate.py:32]: why is it showing a diff here

File: api/translate.py [hunk 12]
```diff
@@ -1,25 +1,35 @@
...
+        message: 'telegram.Message' = update.message
+    else:
+        return
 
+    text: str
     if not context.args:
         try:
             args = message.reply_to_message.text or message.reply_to_message.caption
             translator = Translator()
             translated = translator.translate(args, dest='en')
```
Comment (Tyagdit) [inline api/translate.py]: here 3

File: api/stats.py [hunk 13]
```diff
@@ -1,8 +1,14 @@
...
 from mysql.connector import errorcode
 
 from configuration import config
 
+if TYPE_CHECKING:
+    import telegram
+    import telegram.ext
+
 try:
     conn = mysql.connector.connect(
```
Comment (Tyagdit) [inline api/stats.py:13]: here 4

File: api/search.py [hunk 14]
```diff
@@ -97,23 +98,23 @@ def search(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -
...
-                or result.fileName.startswith("Jackett: api key error!")
+                    # TPB returns a placeholder result if nothing is found
+                    (
+                            result.fileName == "No results returned"
+                            and result.nbSeeders == 0
+                            and result.nbLeechers == 0
+                            and result.fileSize == 0
+                    )
+                    # Jackett error still shows up sometimes
+                    or result.fileName.startswith("Jackett: api key error!")
```
Comment (Tyagdit) [inline api/search.py]: bro wtf fix your indent settings

File: api/response.py [hunk 15]
```diff
@@ -25,9 +27,9 @@ def response(message: Optional['telegram.Message'], action: str):
     if not message:
         return
 
-    response = get(f'https://some-random-api.ml/animu/{action}').json()
+    r = get(f'https://some-random-api.ml/animu/{action}').json()
```
Comment (Tyagdit) [inline api/response.py]: oh i missed this one

File: api/pic.py [hunk 16]
```diff
@@ -1,20 +1,24 @@
...
-    while True:
-        img_code = choices(chars, k=length)
-        address = 'https://i.imgur.com/' + "".join(img_code) + '.jpg'
-        img = Image.open(BytesIO(get(address).content))
+    chars: str = '01234567890ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghiklmnopqrstuvwxyz'
+    length: int = 5
 
-        if img.size[0] != 161:
+    while True:
+        address = 'https://i.imgur.com/' + "".join(choices(chars, k=length)) + '.jpg'
```
Comment (Tyagdit) [inline api/pic.py]: here too

File: api/pfp.py [hunk 17]
```diff
@@ -1,11 +1,20 @@
...
+def pad_image(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -> None:
     """Pad images to 1:1 for use in profile pictures"""
 
-    message = update.message
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
 
     if not (message.reply_to_message and message.reply_to_message.photo):
```
Comment (Tyagdit) [inline api/pfp.py]: in this part too

File: api/joke.py [hunk 18]
```diff
@@ -1,18 +1,25 @@
...
+        response = get('https://sv443.net/jokeapi/v2/joke/Any').json()
     else:
-        response = get('https://official-joke-api.appspot.com/random_joke')
+        response = get('https://official-joke-api.appspot.com/random_joke').json()
 
-    response = response.json()
+    text: str
 
     update.message.reply_text(text=response["setup"])
     time.sleep(2.0)
```
Comment (Tyagdit) [inline api/joke.py:30]: below this

File: api/insult.py [hunk 19]
```diff
@@ -12,7 +14,9 @@ def insult(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -
         message: 'telegram.Message' = update.message
     else:
         return
+
     response: dict = get('https://evilinsult.com/generate_insult.php?lang=en&type=json').json()
+    text: str
```
Comment (Tyagdit) [inline api/insult.py]: this is unused

File: api/hltb.py [hunk 20]
```diff
@@ -1,30 +1,42 @@
...
+
 from howlongtobeatpy import HowLongToBeat
 
+if TYPE_CHECKING:
+    import telegram
+    import telegram.ext
+
+
+def hltb(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -> None:
+    """Query HLTB for a game"""
```
Comment (Tyagdit) [inline api/hltb.py]: i think previous docstring is clearer, HLTB isn't a popular acronym

File: api/hltb.py [hunk 21]
```diff
@@ -1,30 +1,42 @@
...
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
 
-def hltb(update, context):
-    """Find how long a game takes to beat"""
-    message = update.message
-    game = ' '.join(context.args)
+    game = ' '.join(context.args) if context.args else ''
```
Comment (Tyagdit) [inline api/hltb.py]: here too

File: api/hltb.py [hunk 22]
```diff
@@ -1,30 +1,42 @@
...
-        text = "*Usage:* `/hltb {GAME_NAME}`\n"\
+        text = "*Usage:* `/hltb {GAME_NAME}`\n" \
                "*Example:* `/hltb horizon zero dawn`"
     else:
-        results = HowLongToBeat().search(game)
+        results: Optional[List] = HowLongToBeat().search(game, similarity_case_sensitive=False)
 
         if results:
             # Return result with highest similarity to query
             best_guess = max(results, key=lambda element: element.similarity)
```
Comment (Tyagdit) [inline api/hltb.py]: here too

File: api/caption.py [hunk 23]
```diff
@@ -1,27 +1,46 @@
...
+    import telegram.ext
+
 
-def caption(update, context):
+def caption(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -> None:
     """Uses NLP to generate a caption for an image."""
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
```
Comment (Tyagdit) [inline api/caption.py]: same as before, needs to be in the try block

File: api/caption.py [hunk 24]
```diff
@@ -1,27 +1,46 @@
...
     try:
-        message = update.message.reply_to_message
+        message = message.reply_to_message
 
         file = context.bot.getFile(message.photo[-1].file_id)
         configuration = cloudmersive_image_api_client.Configuration()
         configuration.api_key['Apikey'] = config["CLOUDMERSIVE_API_KEY"]
 
-        api_instance = cloudmersive_image_api_client.RecognizeApi(cloudmersive_image_api_client.ApiClient(configuration))
+        api_instance = cloudmersive_image_api_client.RecognizeApi(
```
Comment (Tyagdit) [inline api/caption.py]: missed these

File: api/calc.py [hunk 25]
```diff
@@ -22,9 +25,9 @@ def calc(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') ->
         client: wolframalpha.Client = wolframalpha.Client(config["WOLFRAM_APP_ID"])
         result: wolframalpha.Result = client.query(query)
 
-        if result.success == 'true':
+        try:
             text = next(result.results).text
-        else:
+        except StopIteration:
```
Comment (Tyagdit) [inline api/calc.py]: we should check the success condition along with this exception

File: api/audio.py [hunk 26]
```diff
@@ -31,12 +34,12 @@ def audio(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') ->
...
-        and update.effective_chat
-        and update.effective_chat.type == "private"
-        and message.reply_to_message
-        and message.reply_to_message.voice
+            message.from_user
+            and message.from_user.username in config["AUDIO_RESTORE_USERS"]
+            and update.effective_chat
+            and update.effective_chat.type == "private"
+            and message.reply_to_message
+            and message.reply_to_message.voice
```
Comment (Tyagdit) [inline api/audio.py:46]: fix indent pls

File: api/age.py [hunk 27]
```diff
@@ -1,9 +1,22 @@
...
+def age(update: 'telegram.Update', context: 'telegram.ext.CallbackContext') -> None:
     """Guesses age and gender of people in an image."""
+    if update.message:
+        message: 'telegram.Message' = update.message
+    else:
+        return
+
+    text: str
     try:
         message = update.message.reply_to_message
```
Comment (Tyagdit) [inline api/age.py:21]: here too

**Review by Tyagdit — COMMENTED** [review]
pls do the needful

File: api/wait.py (hunk 7 above)
Comment (obviyus) [reply api/wait.py]: `update.message` might not exist, return immediately. Later down we try for `update.reply_to_message` that may or may not exist.

File: api/translate.py (hunk 11 above)
Comment (obviyus) [reply api/translate.py:32]: Must be this: https://stackoverflow.com/questions/29045140/env-bash-r-no-such-file-or-directory

Hate it, used `dos2unix` to get rid of those but apparently still sneaking in.

File: api/calc.py (hunk 25 above)
Comment (obviyus) [reply api/calc.py]: I don't think the `Result.success` var exists anymore. Couldn't find it in the docs or the source.

File: api/wait.py (hunk 7 above)
Comment (Tyagdit) [reply api/wait.py]: so we check for both in the try block, something like
```suggestion
    if update.message and update.message.reply_to_message:
        message: 'telegram.Message' = update.message.reply_to_message
    else:
        return
```

File: api/calc.py (hunk 25 above)
Comment (Tyagdit) [reply api/calc.py]: weird, check the status code instead then

**Review by obviyus — APPROVED** [review]

**Pull request merged**